library/ssl_cli.c

```diff
@@ -554,31 +553,29 @@ static void ssl_write_extended_ms_ext( mbedtls_ssl_context *ssl,
     *p++ = 0x00;
 
     *olen = 4;
+
+    return( 0 );
 }
 #endif /* MBEDTLS_SSL_EXTENDED_MASTER_SECRET */
 
 #if defined(MBEDTLS_SSL_SESSION_TICKETS)
-static void ssl_write_session_ticket_ext( mbedtls_ssl_context *ssl,
-                                          unsigned char *buf, size_t *olen )
+static int ssl_write_session_ticket_ext( mbedtls_ssl_context *ssl,
+					 unsigned char *buf,
+					 const unsigned char *end,
+					 size_t *olen )
 {
     unsigned char *p = buf;
-    const unsigned char *end = ssl->out_msg + MBEDTLS_SSL_MAX_CONTENT_LEN;
     size_t tlen = ssl->session_negotiate->ticket_len;
 
     *olen = 0;
 
     if( ssl->conf->session_tickets == MBEDTLS_SSL_SESSION_TICKETS_DISABLED )
-    {
-        return;
-    }
+        return( 0 );
 
     MBEDTLS_SSL_DEBUG_MSG( 3, ( "client hello, adding session ticket extension" ) );
 
-    if( end < p || (size_t)( end - p ) < 4 + tlen )
-    {
-        MBEDTLS_SSL_DEBUG_MSG( 1, ( "buffer too small" ) );
-        return;
-    }
+    /* The addition is safe here since the ticket length is 16 bit. */
+    MBEDTLS_SSL_CHK_BUF_PTR_WITH_DEBUG( p, end, 4 + tlen );
```

Comment:
@andresag01 @hanno-arm You both make good points here. Technically `end - 4` could overflow; it isn't actually going to happen with 4 but it would happen for a large enough 4 and I don't want to have to think about what the maximum subtractable constant is. But the parts of the code affected by this PR are ok (except on 16-bit platforms, but we have code in the library that wouldn't even compile on a 16-bit platform) so I side with deferring until later. We're tracking this internally as IOTSSL-1885.